### PR TITLE
npiet 1.3e (new formula)

### PR DESCRIPTION
**Pull request by habibalamin**
I had to change `<malloc.h>` inclusions in some source files to `<malloc/malloc.h>`. `malloc` should be accessible from `<stdlib.h>` (as per the C89 standard)<sup>[1](http://earthworm.isti.com/trac/earthworm/ticket/141)</sup>, so that should technically be changed upstream. I'm just about to contact the maintainer, Erik Schoenfelder.

Not sure if this falls under the notable banner, but it's an implementation of the esoteric Piet programming language, which made Guy Steele's & Richard Gabriel's “"50 in 50" talk about the history of computer languages, which they have presented at several major computer science conferences”. The homepage for Piet mentions implementations around the web and notes this one.

### CODE REVIEW

**Comment by habibalamin** [comment]
Ah, I should probably explain why I use `IO.popen`, and that's because there's no `--help` option for `npiet`; you simply have to misuse it to get the help message printed to `stderr`. If you like, I can add a comment to the formula explaining that.

**Comment by habibalamin** [comment]
This formula could be improved by making each image format support optional (with a default), so you don't have to link against unnecessary libraries, for instance, if you only want GIF support.

File: Formula/npiet.rb
Comment (MikeMcQuaid) [inline Formula/npiet.rb]: Is there an upstream issue tracking progress on using the most recent version?

File: Formula/npiet.rb
Comment (MikeMcQuaid) [inline Formula/npiet.rb]: Can you please submit these `inreplace` calls as patches to the upstream developers of this project and add a link to the upstream patch submissions and explanation of why the patches are needed in a comment in the formula file. Thanks!

File: Formula/npiet.rb
Comment (MikeMcQuaid) [inline Formula/npiet.rb]: Can `shell_output` be used here with e.g. `2>&1`?

File: Formula/npiet.rb
Comment (MikeMcQuaid) [inline Formula/npiet.rb]: Flip the `assert_match`es; the expected/actual are the wrong way around.

File: Formula/npiet.rb
Comment (MikeMcQuaid) [inline Formula/npiet.rb]: Just inline this value.

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]

File: Formula/npiet.rb
Comment (habibalamin) [inline Formula/npiet.rb]: 5 is the latest. A lot of Homebrew formulae have not been updated to work with 5, however, so Homebrew keeps core as version 4.

File: Formula/npiet.rb
Comment (habibalamin) [inline Formula/npiet.rb]: The `Makefile` is generated by the `configure` script. I'll report that the `--mandir` option doesn't seem to be working upstream, but it's not just a simple patch to the `Makefile`. It has to be patched after `./configure` is run.

File: Formula/npiet.rb
Comment (habibalamin) [inline Formula/npiet.rb]: It's used in multiple methods, and multiple times in each.

File: Formula/npiet.rb
Comment (habibalamin) [inline Formula/npiet.rb]: Will do. I wanted it to line up without having large gaps for the PNG and PPM outputs compared to GIF, but it's not strictly necessary.

File: Formula/npiet.rb
Comment (habibalamin) [inline Formula/npiet.rb]: Will do. I was trying to keep the shell-isms low, but you're right, that way is better.

File: Formula/npiet.rb
Comment (MikeMcQuaid) [inline Formula/npiet.rb]: Ok, as long as we're reporting this that'd be good.

File: Formula/npiet.rb
Comment (MikeMcQuaid) [inline Formula/npiet.rb]: Sure, we don't need formulae to be DRY.

File: Formula/npiet.rb
Comment (MikeMcQuaid) [inline Formula/npiet.rb]: Don't bother aligning these.

File: Formula/npiet.rb
Comment (MikeMcQuaid) [inline Formula/npiet.rb]: Can skip the `.split("\n").first`; it'll match anyway.

**Comment by habibalamin** [comment]
@MikeMcQuaid done.

**Comment by MikeMcQuaid** [comment]
@habibalamin This will need to wait until the newer giflib is part of homebrew/core. You could wait for me to do that or give it a go yourself (which I'm happy to help with).

**Comment by MikeMcQuaid** [comment]
@habibalamin Also: still waiting for upstream patch submissions/links. Thanks!

**Comment by habibalamin** [comment]
@MikeMcQuaid I've sent an email to the upstream author/maintainer of `npiet` with a patch.

It looks like it might be better to remove the `<malloc.h>` includes entirely, as they're already included as part of `<stdlib.h>`, as mentioned in my first comment on this PR, which I realised later *has* already been included.

I'll leave the patchfile here, so you can take a look. I still keep one `<malloc.h>` that is included for Visual C++ (in an `#ifdef` chain), but is local to some code that I think is Bison-related and I am not willing to touch; I'm not a C programmer, nor a PL author/user of tools like Bison.

[macos-malloc-fix.patch.zip](https://github.com/Homebrew/homebrew-core/files/772430/macos-malloc-fix.patch.zip)

As for `--mandir` not working, not sure what's up with that. It could just be a `configure` program that has a generic help message, but doesn't support all the options. I couldn't say. In any case, I've let upstream know.

**Comment by MikeMcQuaid** [comment]
> @MikeMcQuaid I've sent an email to the upstream author/maintainer of npiet with a patch.
> As for --mandir not working, not sure what's up with that. It could just be a configure program that has a generic help message, but doesn't support all the options. I couldn't say. In any case, I've let upstream know.

Can you note these as Ruby comments in the formula ("add a link to the upstream patch submission and explanation of why the patch is needed in a comment in the formula file")? 

Also, let me know whether you're interested in moving the `giflib5` formula to `giflib@5` in this tap; if not I'll help.

Thanks!

**Comment by habibalamin** [comment]
I guess I'm not sure what you mean by upstream patch submission. I don't have a link to that. I sent a patchfile by email to the upstream maintainer.

I'm too busy to help with the `giflib` stuff.

**Comment by MikeMcQuaid** [comment]
> I guess I'm not sure what you mean by upstream patch submission. I don't have a link to that. I sent a patchfile by email to the upstream maintainer.

Can you detail that in the formula? Thanks.

> I'm too busy to help with the giflib stuff.

No worries, I'll merge this after I've got around to it.

**Comment by habibalamin** [comment]
Can we open this again, since giflib5 has been added to the core tap?

**Comment by MikeMcQuaid** [comment]
@habibalamin feel free to reopen a new PR based on this.

**Pull request closed without merging**
